### PR TITLE
Type alias/member ordering

### PR DESCRIPTION
**Pull request by elliottt**
The resolver was handling type aliases before type members, which in some cases would cause the bounds of a type member to be fixed as `T.untyped` in a call to `externalType`. This PR fixes that by tracking ordering dependencies between aliases and type members, and resolving them only once their dependencies have been resolved.

### Motivation

This bug was blocking the attached class PR #1855, as it adds a type member to every singleton class.

### Test plan

Added and updated some tests.

See included automated tests.

### CODE REVIEW

File: resolver/resolver.cc [hunk 1]
```diff
@@ -847,118 +844,333 @@ class ResolveConstantsWalk {
...
+    struct ResolveAssignItem {
+        // The symbol being populated, either a type alias or an individual
+        // type_member.
+        core::SymbolRef lhs;
+
+        // The type_member, type_template, or type_alias of the RHS.
+        ast::Send *rhs;
+
+        // The symbols that this alias depends on that have type members.
+        vector<core::SymbolRef> dependencies;
```
[the remark this answers is not in the document]
Comment (jez) [reply resolver/resolver.cc]: from irl: it's nicer to keep it the way it is

File: test/testdata/resolver/type_member_cycle.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  X = type_member(fixed: B)
+# ^^^^^^^^^^^^^^^^^^^^^^^^^ error: Type member `A::X` is involved in a cycle
+end
+
+class B
+  extend T::Sig
+  extend T::Generic
+
+  X = type_member(fixed: A)
+# ^^^^^^^^^^^^^^^^^^^^^^^^^ error: Type member `B::X` is involved in a cycle
```
Comment (jez) [inline test/testdata/resolver/type_member_cycle.rb]: Nice I like that we report the error in both places.

File: test/testdata/resolver/type_member_out_of_order.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+  X = type_member(fixed: Integer)
+
+  sig {returns(X)}
+  def test
+    10
+  end
+end
+
+T.reveal_type(B.new.test) # error: Revealed type: `Integer`
```
Comment (jez) [inline test/testdata/resolver/type_member_out_of_order.rb]: 👍

File: test/testdata/infer/generics/bounds.rb [hunk 4]
```diff
@@ -1,4 +1,5 @@
 # typed: true
+# disable-fast-path: true
```
[the remark this answers is not in the document]
Comment (elliottt) [reply test/testdata/infer/generics/bounds.rb]: The logic here will exit early if the LHS has already been resolved, and as a result the type errors don't get re-raised in the fast path.

File: resolver/resolver.cc [hunk 5]
```diff
@@ -847,118 +844,333 @@ class ResolveConstantsWalk {
...
+    // definition
+    bool trackDependencies_ = false;
+    vector<bool> classOfDepth_;
+    vector<core::SymbolRef> dependencies_;
+
+    bool isInsideClassOf() const {
+        return !classOfDepth_.empty() && classOfDepth_.back();
+    }
+
+    void clearTypeContext(bool tracking) {}
```
[the remark this answers is not in the document]
Comment (elliottt) [reply resolver/resolver.cc]: Yep, thanks for catching this! It was originally for managing the `trackDependencies_`, `dependencies_` and `classOfDepth_` state, but it ended up only being used once.

File: resolver/resolver.cc [hunk 6]
```diff
@@ -847,118 +844,333 @@ class ResolveConstantsWalk {
...
-                    if (auto e = ctx.state.beginError(send->loc, core::errors::Resolver::ParentTypeBoundsMismatch)) {
-                        e.setHeader("parent lower bound `{}` is not a subtype of lower bound `{}`",
-                                    parentType->lowerBound->show(ctx), memberType->lowerBound->show(ctx));
-                    }
+            case core::Names::typeAlias()._id:
+                resolveTypeAlias(ctx, job.lhs, job.rhs);
+                break;
+        }
+
+        return true;
```
[the remark this answers is not in the document]
Comment (elliottt) [reply resolver/resolver.cc]: it should be an enforce failure to make it this far without having the RHS be `T.type_alias`, `type_member`, or `type_template`, so maybe I could just rephrase this as an `if` as that would read a little better.

**Comment by elliottt** [comment]
Pay-server build: http://go/caGArVsS

**Comment by jez** [comment]
New changes lgtm
